provider/transip.go

```diff
@@ -0,0 +1,373 @@
+		if err := transip.SetDNSEntries(p.client, zone.Name, zone.DNSEntries); err != nil {
+			log.WithFields(log.Fields{"zone": zone.Name, "error": err.Error()}).Warn("failed to update")
+		}
+	}
+
+	return nil
+}
+
+// fetchZones returns a list of all domains within the account
+func (p *TransIPProvider) fetchZones() ([]transip.Domain, error) {
+	domainNames, err := transip.GetDomainNames(p.client)
+	if err != nil {
+		return nil, err
+	}
+
+	domains, err := transip.BatchGetInfo(p.client, domainNames)
+	if err != nil {
+		return nil, err
+	}
+
+	var zones []transip.Domain
+	for _, d := range domains {
+		if !p.domainFilter.Match(d.Name) {
+			continue
+		}
+
+		zones = append(zones, d)
+	}
+
+	return zones, nil
+}
+
+// Zones returns the list of hosted zones.
+func (p *TransIPProvider) Zones() ([]transip.Domain, error) {
+	zones, err := p.fetchZones()
+	if err != nil {
+		return nil, err
+	}
+
+	return zones, nil
+}
+
+// Records returns the list of records in a given zone.
+func (p *TransIPProvider) Records() ([]*endpoint.Endpoint, error) {
+	zones, err := p.Zones()
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []*endpoint.Endpoint
+	var name string
+	// go over all zones and their DNS entries and create endpoints for them
+	for _, zone := range zones {
+		for _, r := range zone.DNSEntries {
+			if !supportedRecordType(string(r.Type)) {
+				continue
+			}
+
+			name = p.endpointNameForRecord(r, zone)
+			endpoints = append(endpoints, endpoint.NewEndpointWithTTL(name, string(r.Type), endpoint.TTL(r.TTL), r.Content))
+		}
+	}
+
+	return endpoints, nil
+}
+
+// endpointNameForRecord returns "www.example.org" for DNSEntry with Name "www" and
+// Doman with Name "example.org"
+func (p *TransIPProvider) endpointNameForRecord(r transip.DNSEntry, d transip.Domain) string {
+	// root name is identified by "@" and should be translated to domain name for
+	// the endpoint entry.
+	if r.Name == "@" {
+		return d.Name
+	}
+
+	return fmt.Sprintf("%s.%s", r.Name, d.Name)
+}
+
+// recordNameForEndpoint returns "www" for Endpoint with DNSName "www.example.org"
+// and Domain with Name "example.org"
+func (p *TransIPProvider) recordNameForEndpoint(ep *endpoint.Endpoint, d transip.Domain) string {
+	// root name is identified by "@" and should be translated to domain name for
+	// the endpoint entry.
+	if ep.DNSName == d.Name {
+		return "@"
+	}
+
+	return strings.TrimSuffix(ep.DNSName, "."+d.Name)
+}
+
+// getMinimalValidTTL returns max between given Endpoint's RecordTTL and
+// transipMinimalValidTTL
+func (p *TransIPProvider) getMinimalValidTTL(ep *endpoint.Endpoint) int64 {
+	// TTL cannot be lower than transipMinimalValidTTL
+	if ep.RecordTTL < transipMinimalValidTTL {
+		return transipMinimalValidTTL
+	}
+
+	return int64(ep.RecordTTL)
+}
+
+// dnsEntriesAreEqual compares the entries in 2 sets and returns true if the
+// content of the entries is equal
+func (p *TransIPProvider) dnsEntriesAreEqual(a, b transip.DNSEntries) bool {
```

Comment:
I'm not sure if we already have a function like this. Did you check the code from other providers if there's anything related?

Reply:
I'm specifically comparing 2 objects with types from the TransIP API. I certainly looked around how other providers apply zone changes, couldn't find something similar.